mining: drop unused  -nFees and sigops from CBlockTemplate

Sjors opened the pull request:
For the coinbase `vTxFees` used a dummy value of -nFees.

Similarly the first `vTxSigOpsCost` entry was calculated from
the dummy coinbase transaction.

This was introduced in #2115, but the values were never returned by the RPC or used in a test.

Drop 'm and add code comments to prevent confusion.

This PR also adds test coverage for the `fees` and `sigops` fields in `getblocktemplate`, so it closes #32053.

Sjors commented on src/node/miner.cpp:
```diff
@@ -124,10 +124,12 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
     pblock->vtx.emplace_back();
-    pblocktemplate->vTxFees.push_back(-1); // updated at end
-    pblocktemplate->vTxSigOpsCost.push_back(-1); // updated at end
+    pblocktemplate->vTxFees.push_back(0);
```
Changed from -1 to 0, because if someone ignores the repeated instruction to not use this value, at least they won't incorrectly sum up `vTxFees`.

ryanofsky reviewed (approved):
Code review ACK d1cc874d1f87a654aa2da50dbcf850b84517e938

This does seem like a good simplification, but maybe instead of changing from non-zero dummy values to zero dummy values, it could make sense to drop dummy values entirely like:

<details><summary>diff</summary>
<p>

```diff
--- a/src/interfaces/mining.h
+++ b/src/interfaces/mining.h
@@ -37,11 +37,9 @@ public:
     // Block contains a dummy coinbase transaction that should not be used.
     virtual CBlock getBlock() = 0;
 
-    // Fees per transaction. Position 0 represents a dummy coinbase that should
-    // not be used.
+    // Fees per transaction, not including coinbase transaction.
     virtual std::vector<CAmount> getTxFees() = 0;
-    // Sigop cost per transaction. Position 0 represents a dummy coinbase that should
-    // not be used.
+    // Sigop cost per transaction, not including coinbase transaction.
     virtual std::vector<int64_t> getTxSigops() = 0;
 
     virtual CTransactionRef getCoinbaseTx() = 0;
--- a/src/node/miner.cpp
+++ b/src/node/miner.cpp
@@ -128,8 +128,6 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
     // sigops to it. The coinbase is skipped by the getblocktemplate RPC.
     // Mining interface consumers must not use it.
     pblock->vtx.emplace_back();
-    pblocktemplate->vTxFees.push_back(0);
-    pblocktemplate->vTxSigOpsCost.push_back(0);
 
     LOCK(::cs_main);
     CBlockIndex* pindexPrev = m_chainstate.m_chain.Tip();
@@ -175,7 +173,6 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
     UpdateTime(pblock, chainparams.GetConsensus(), pindexPrev);
     pblock->nBits          = GetNextWorkRequired(pindexPrev, pblock, chainparams.GetConsensus());
     pblock->nNonce         = 0;
-    pblocktemplate->vTxSigOpsCost[0] = WITNESS_SCALE_FACTOR * GetLegacySigOpCount(*pblock->vtx[0]);
 
     BlockValidationState state;
     if (m_options.test_block_validity && !TestBlockValidity(state, chainparams, m_chainstate, *pblock, pindexPrev,
--- a/src/node/miner.h
+++ b/src/node/miner.h
@@ -37,9 +37,9 @@ static const bool DEFAULT_PRINT_MODIFIED_FEE = false;
 struct CBlockTemplate
 {
     CBlock block;
-    // Fees per transaction. Position 0 represents a dummy coinbase.
+    // Fees per transaction, not including coinbase transaction.
     std::vector<CAmount> vTxFees;
-    // Sigops per transaction. Position 0 represents a dummy coinbase.
+    // Sigops per transaction, not including coinbase transaction.
     std::vector<int64_t> vTxSigOpsCost;
     std::vector<unsigned char> vchCoinbaseCommitment;
     /* A vector of package fee rates, ordered by the sequence in which
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -893,7 +893,7 @@ static RPCHelpMan getblocktemplate()
         }
         entry.pushKV("depends", std::move(deps));
 
-        int index_in_template = i - 1;
+        int index_in_template = i - 2;
         entry.pushKV("fee", tx_fees.at(index_in_template));
         int64_t nTxSigOps{tx_sigops.at(index_in_template)};
         if (fPreSegWit) {
```
</p>
</details>

Also if clients of IPC interface just need total fees and sigops, it might make sense to a add a method returning both of those totals to make interface more convenient and efficient and not require multiple calls.

ismaelsadeeq commented:
Concept ACK
Will ack after https://github.com/bitcoin/bitcoin/pull/31897#pullrequestreview-2624259635

Sjors commented:
> maybe instead of changing from non-zero dummy values to zero dummy values, it could make sense to drop dummy values entirely

I think it's more intuitive for `block.vtx` and these two vectors to have the same length.

> Also if clients of IPC interface just need total fees and sigops

Agreed, but at this point I haven't seen a need for it yet.

ismaelsadeeq commented:
ACK d1cc874d1f87a654aa2da50dbcf850b84517e938

> I think it's more intuitive for block.vtx and these two vectors to have the same length.

No strong opinion but I tend to prefer @ryanofsky  diff because it is precise. Since the coinbase transaction does not have fees, we should not simply add 0 for the size of `vTxFees` to match `vtx` .

They are two different list with different meaning
`vtx` is the list of transactions in the block, while `vTxFees` is the list of fees for each transaction in the block which should exclude the coinbase.

For `vTxSigOpsCost` I agree it should match `vtx` and 0 is okay because eventually there will be coinbase but we should clearly indicate the 0 there is  dummy.


This is also similar to what was done in https://github.com/bitcoin/bitcoin/pull/31384/commits/777434a2cd14841e35ce39d7a6f51131e6a41de2 to improve precision of the block assembler.

Sjors commented on src/node/miner.cpp:
```diff
@@ -174,7 +175,6 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
     UpdateTime(pblock, chainparams.GetConsensus(), pindexPrev);
     pblock->nBits          = GetNextWorkRequired(pindexPrev, pblock, chainparams.GetConsensus());
     pblock->nNonce         = 0;
-    pblocktemplate->vTxSigOpsCost[0] = WITNESS_SCALE_FACTOR * GetLegacySigOpCount(*pblock->vtx[0]);
```
I forgot to drop this line in my previous push.

Sjors commented:
`block.vtx` contains a dummy coinbase, `vTxFees` and `vTxSigOpsCost` contain dummy fees and dummy sigop costs for the coinbase. That seems more straight forward to me than having `vTxFees` be shorter and then having to document why these things have different lengths.

> Since the coinbase transaction does not have fees

That's almost a philosophical question, because miners can (and sometimes accidentally did) claim less fees than allowed. In a way such a coinbase has negative fees, which is perhaps why the original code did that.

Perhaps the most rigorous solution is to add `std::optional<CAmount> fee` to `CTransaction`, but that would have implications all over the codebase, so I'd rather not go on that adventure.

ismaelsadeeq commented:
> block.vtx contains a dummy coinbase, vTxFees and vTxSigOpsCost contain dummy fees and dummy sigop cost

Good point.

> That seems more straight forward to me than having vTxFees be shorter and then having to document why these things have different lengths.


We don't have to document that `vTxFees` is just a vector of fees paid in the block.

Note: This is an unblocking comment.

This PR is an immediate win because we don't have to perform checks like we do in https://github.com/bitcoin/bitcoin/pull/31283/commits/d4020f502a63cb4390ec241fc5f989e988afa022#diff-0ef8ae12c6e9ef2accc78537f42612b3267e8a7c45dc7e9eb998f797e79f2e95R979-R980, should this PR come before #31283?

Sjors commented:
> This PR is an immediate win because we don't have to perform checks like we do in

No this PR doesn't change anything there. It's just that people suggested using `-vTxFees[]` directly there instead of doing the calculation. This PR removes that temptation (and makes it easier to get rid of `vTxFees` entirely at some point).

ryanofsky reviewed (approved):
Code review ACK 587ad67bf3bee03a7088e1dd5be2828b7a4f4fd4. Only change since last review is removing another dummy vTxSigOpsCost assignment.

IMO, it would be probably be more ideal to drop the dummy values from the `CBlockTemplate::vTxFees` and `CBlockTemplate::vTxSigOpsCost` vectors. I don't think it matters if they have different indexing than `CBlock::vtx` because that is a different vector in a different class, and it would be confusing for that vector *not* to contain a placeholder transaction, because it would make indexing in the `CBlock` unstable over time and updating the `CBlock` inefficient. These reasons seem inapplicable to the `CBlockTemplate` members. I think the documentation in [suggested diff](https://github.com/bitcoin/bitcoin/pull/31897#pullrequestreview-2624259635) was sufficient but could be updated with

```diff
-// Fees per transaction, not including coinbase transaction.
+// Fees per transaction, not including coinbase transaction (unlike CBlock::vtx)
 std::vector<CAmount> vTxFees;
```

if an explicit comparison would be helpful. No strong opinion on this though, just explaining my thinking.

Sjors committed:
mining: drop unused -nFees and sigops from CBlockTemplate

For the coinbase vTxFees used a dummy value of -nFees. This
value was never returned by the RPC or used in a test.

Similarly the fist vTxSigOpsCost entry was calculated from
the dummy coinbase transaction.

Drop both and add code comments to prevent confusion.

Sjors commented:
Rebased and dropped coinbase from `vTxFees` and `vTxSigOpsCost`.

luke-jr commented:
Concept NACK, just makes the code less straightforward and forces re-calculation later if needed. Harmless to leave it, and make use  (eg, in #31283 and other feature I haven't PR'd yet)

ryanofsky reviewed (approved):
Code review ACK 6d7648795aa053f535510d11379fdd9d0399004c. Since last review, this was changed to drop dummy values instead of setting them to 0.

re: https://github.com/bitcoin/bitcoin/pull/31897#issuecomment-2683554828

> Concept NACK, just makes the code less straightforward and forces re-calculation later if needed. Harmless to leave it

I'm confused about why total fees were stored in a dummy field, and why the fields were represented as a negative number, and why the field didn't seem to be documented and also didn't seem to be used. But I also don't have a clear idea of why individual transaction fees as opposed to total fees are useful here either, so it is hard for me to form a real opinion other than that this change seems like a simplification and improvement that could be prevent bugs from using the undocumented dummy values incorrectly.

> and make use (eg, in #31283 and other feature I haven't PR'd yet)

IMO it doesn't simplify #31283 to use `-vTxFees[0]` instead of adding up the fees, but it could be nice to add a separate `total_fees` field. Maybe `total_fees` could be useful for the other feature not PRed yet too?

Sjors commented:
> forces re-calculation later if needed

It wouldn't. We can introduce `total_fees` and set it in the same place where we now set this dummy value.

The recalculation in #31283 is done in order to avoid relying on this code. I considered adding `total_fees` for that PR, but it doesn't seem useful because I intend to replace that code when cluster mempool comes around.

> other feature I haven't PR'd yet

That would justify a `total_fees` field. I'm happy to refactor `waitNext()` after said PR is merged, if that happens before the cluster mempool changes.

ismaelsadeeq reviewed (approved):
Code review ACK 6d7648795aa053f535510d11379fdd9d0399004c  
 
This dummy values were introduced in #2115. I did not see the usage or rationale for adding the negative fees of all block transactions as the fee of the coinbase tx.
I think this dummy values should be deleted.

However I think having total_fees values can be useful.

glozow reviewed (commented):
looks correct, untested ACK 6d7648795aa053f535510d11379fdd9d0399004c

Sjors commented on src/rpc/mining.cpp:
```diff
@@ -893,7 +893,7 @@ static RPCHelpMan getblocktemplate()
         entry.pushKV("depends", std::move(deps));
 
-        int index_in_template = i - 1;
+        int index_in_template = i - 2;
```
Tracking in #32053.

Sjors replied:
Pushed a test.

Sjors committed:
test: check fees and sigops in getblocktemplate

Sjors commented:
Rebased and switched the order of commits, so it's more clear this PR doesn't change behavior.

Sjors commented on test/functional/mining_basic.py:
```diff
@@ -93,6 +96,57 @@ def mine_chain(self):
+
+        # Generate transaction without sigops. It will go first because it pays
+        # higher fees (100 sat/vbyte) and descends from a different coinbase.
+        tx_d = self.wallet.send_self_transfer(from_node=node,
+                                              fee_rate=Decimal("0.00100"))
```
The two different `MiniWallet` instances can't spend each others coins, so I don't think these additions are needed - but useful to keep in mind.

ismaelsadeeq commented:
re-ACK 226d81f8b708ea1c3a39ec58446e291fe7440fdd 
 
Latest change added functional test coverage for the `sigops` and `fee` field of txs returned from `getblocktemplate` RPC

ryanofsky reviewed (approved):
Code review ACK 226d81f8b708ea1c3a39ec58446e291fe7440fdd. New test was added since last review, which seems very cleanly written and fixes some missing coverage.

glozow reviewed (commented):
ACK 226d81f8b708ea1c3a39ec58446e291fe7440fdd

Thanks for adding the test! Checked that it'd catch any errors in the RPC code.
